### ISSUE TITLE
EF Interceptor should not crash the workflow if something fails during audit

### ISSUE THREAD

**Issue by lucax88x** [thepirat000/Audit.NET#589]
**Describe the bug**
Right now I have two problems:
1) EF interceptor is crashing, and I need to analyze why it's crashing
2) BUT, when EF interceptor crashes, the workflow stops, IMHO audit should be "invisible", what do you think?

**To Reproduce**
- add AuditSaveChangesInterceptor
- method that creates dbContext using injected dbcontextpooledfactory and when it save it will fail

```csharp
        await using var dbContext = await _stammdatenDbContextFactory.CreateDbContextAsync();
       // do stuff
       await dbContext.SaveChangesAsync(); <-- save fails for exceptions
```

```exception
System.ObjectDisposedException: Cannot access a disposed context instance. A common cause of this error is disposing a context instance that was resolved from dependency injection and then later trying to use the same context instance elsewhere in your application. This may occur if you are calling 'Dispose' on the context instance, or wrapping it in a using statement. If you are using dependency injection, you should let the dependency injection container take care of disposing context instances.
Object name: 'StammdatenDbContext'.
   at Microsoft.EntityFrameworkCore.DbContext.CheckDisposed()   at Microsoft.EntityFrameworkCore.DbContext.get_Database()   at Audit.EntityFramework.DbContextHelper.IsRelational(DbContext dbContext)   at Audit.EntityFramework.DbContextHelper.GetClientConnectionId(DbContext dbContext)   at Audit.EntityFramework.DbContextHelper.UpdateAuditEvent(EntityFrameworkEvent efEvent, IAuditDbContext context)   at Audit.EntityFramework.DbContextHelper.SaveScopeAsync(IAuditDbContext context, IAuditScope scope, EntityFrameworkEvent event)   at Audit.EntityFramework.DbContextHelper.EndSaveChangesAsync(IAuditDbContext context, IAuditScope scope, Int32 result, Exception exception)   at Audit.EntityFramework.AuditSaveChangesInterceptor.SaveChangesFailedAsync(DbContextErrorEventData eventData, CancellationToken cancellationToken)   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
```

**Expected behavior**
1) It should not fail
2) it should not crash the workflow when it fails

**Libraries (specify the Audit.NET extensions being used including version):**
 - Audit.EntityFramework.Core: 20.1.3

**Target .NET framework:**
 - .NET 6

**Additional context**

**Comment by thepirat000** [comment]
Please share your configuration (`Audit.Core.Configuration.*` and `Audit.EntityFramework.Configuration.*` calls)

**Comment by thepirat000** [comment]
From the stack trace I can see you're using the EF data provider to save the audits and you save the audits within the same DbContext. So you're probably letting the default behavior of using the same `DbContext` instance for saving the audits.

But whenever the `SaveChanges` fails, the `DbContext` instance will be set to a faulted state or be disposed, so that's why the audit throws the exception.

You could change the EF Data Provider configuration to use a different instance of your DbContext, so any failure in the save operation will not affect the audit. For this, you must provide a way to instantiate your audit DbContext. 

Some examples:

```c#
// Resolve using a parameterless constructor 
Audit.Core.Configuration.Setup()
    .UseEntityFramework(ef => ef
        .UseDbContext<StammdatenDbContext>()
        ...

// Resolve from IServiceProvider (i.e. your project is ASP.NET core MVC)
Audit.Core.Configuration.Setup()
    .UseEntityFramework(ef => ef
        .UseDbContext(ev => ev.EntityFrameworkEvent.GetDbContext().GetService<StammdatenDbContext>())
        ...
```

**Comment by lucax88x** [comment]
I'm not using UseEntityFramework extension, but UsePostgreSql extension instead, so my expectation is that it should use the npgsql driver to write audit stuffs and not the dbcontext audit. Does it make sense what I'm saying?

My configuration:
```csharp
            Configuration
                .Setup()
                .UsePostgreSql(
                    config => config
                        .ConnectionString(dbConnectionString)
                        .Schema(Schema)
                        .TableName(Table)
                        .IdColumnName(IdColumn)
                        .DataColumn(DataColumn, DataType.JSONB)
                        .LastUpdatedColumnName(UpdatedDateColumn)
                );
```

how I add the interceptor:
```csharp

    public static DbContextOptionsBuilder AddMyAuditTrail(this DbContextOptionsBuilder dbContextOptionsBuilder) =>
        dbContextOptionsBuilder.AddInterceptors(new AuditSaveChangesInterceptor());

// somewhere else
        services
            .AddPooledDbContextFactory<TD>(
                builder => {
                    if (addAuditTrail) {
                        builder.AddMyAuditTrail();
                    }

                    builder.UseNpgsql(
                            configuration.GetConnectionString(connectionStringKey),
                            optionsBuilder => {
                                optionsBuilder.EnableRetryOnFailure();
                            }
                        )
                        .UseLowerCaseNamingConvention();
                }
            );


```

How I create tables (by hosted service):

```csharp
    public static async Task CreateTables(IConfiguration configuration) {
        var dbConnectionString = configuration.GetConnectionString(ConnectionStringKey);

        if (!string.IsNullOrEmpty(dbConnectionString)) {
            await using var connection = new NpgsqlConnection(dbConnectionString);
            await connection.OpenAsync();
            
            await using var createSchemaCommand = new NpgsqlCommand(
                $"CREATE SCHEMA IF NOT EXISTS {Schema}",
                connection
            );
            await createSchemaCommand.ExecuteNonQueryAsync();

            await using var createTableCommand = new NpgsqlCommand(
                @$"CREATE TABLE IF NOT EXISTS {Schema}.{Table} (
    {IdColumn} bigserial NOT NULL,
    {InsertedDateColumn} timestamp without time zone NOT NULL DEFAULT now(),
    {UpdatedDateColumn} timestamp without time zone NOT NULL DEFAULT now(),
    {DataColumn} jsonb NOT NULL,
    CONSTRAINT {Table}_pkey PRIMARY KEY (id) 
)",
                connection
            );
            await createTableCommand.ExecuteNonQueryAsync();
        }
        else {
            Log.Warning("No AuditTrail connection string found, not using any audit trail!");
        }
    }
```

**Comment by thepirat000** [comment]
You're right, in that case, I think the audit should not throw.

I would work on the fix to allow continuing with the audit process even if the DbContext got disposed of within the SaveChanges operation.

**Comment by thepirat000** [comment]
This was fixed in version 20.2.4, please upgrade your references and re-test
